### PR TITLE
CFE-3569: Added policy variable sys.os_version_major

### PR DESCRIPTION
**Pull request by larsewi**
Ticket: CFE-3569
Changelog: Title
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

### CODE REVIEW

**Comment by larsewi** [comment]
@cf-bottom Jenkins please

File: tests/acceptance/01_vars/01_basic/os_version_major.cf [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    hpux_11::
+      "expected" string => "11";
+
+    suse_15::
+      "expected" string => "15";
+
+    windows_2012::
+      "expected" string => "2012";
+    windows_2016::
+      "expected" string => "2017";
```
Comment (larsewi) [inline tests/acceptance/01_vars/01_basic/os_version_major.cf:69]: 2016

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6177/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6177/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6177/

**Comment by larsewi** [comment]
`hpux_10` & `hpux_11` are not yet defined on HP-UX'es. But this will be covered in ticket [CFE-3609](https://tracker.mender.io/browse/CFE-3609).

File: libenv/sysinfo.c [hunk 2]
```diff
@@ -3335,6 +3335,46 @@ void GetDefVars(EvalContext *ctx)
...
+                                                             "flavor");
+
+    int major = -1;
+    int minor = -1;
+    char strmajor[PRINTSIZE(major)];
+
+    sscanf(flavor, "%*[^0123456789]%d%*[^0123456789]%d", &major, &minor);
+    if (major != -1)
+    {
+        xsnprintf(strmajor, sizeof(strmajor), "%d", major);
```
Comment (vpodzime) [inline libenv/sysinfo.c:3354]: I don't like this number parsing only to create a string from it again. It's not so hard to identify a start and an end of an integer in a string. And you would not have to worry about `PRINTSIZE()`. Just create a copy of `flavor` so that it can be modified, split it in-place and use it. It's not like we need to optimize this, but we want to look like professionals. :wink:

File: libenv/sysinfo.c [hunk 3]
```diff
@@ -3335,6 +3335,46 @@ void GetDefVars(EvalContext *ctx)
...
         CF_DATA_TYPE_STRING, "invocation,source=agent,command_name=jq");
 }
+
+/*****************************************************************************/
+
+void SysOsVersionMajor(EvalContext *ctx)
+{
+    const char *flavor = EvalContextVariableGetSpecialString(ctx,
+                                                             SPECIAL_SCOPE_SYS,
+                                                             "flavor");
```
Comment (vpodzime) [inline libenv/sysinfo.c:3345]: Please make sure this note from *EvalContextVariableGet()* is addressed:
```
 * @NOTE NULL is a valid return value if #type_out is of list type and the                                                                                                                         
 *       list is empty. To check if the variable didn't resolve, check if                                                                                                                          
 *       #type_out was set to CF_DATA_TYPE_NONE.                                                                                                                                                   
```

**Review by vpodzime — COMMENTED** [review]
Looks good to me otherwise. We cannot merge this with a failing test though.

**Comment by larsewi** [comment]
@cf-bottom if you're not to busy doing bot stuff, could you please run jenkins with exotics ???

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6432/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6432/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6432/

File: libenv/sysinfo.c [hunk 4]
```diff
@@ -3465,6 +3465,69 @@ static void SysOSNameHuman(EvalContext *ctx)
...
+        }
+    }
+
+    if (StringStartsWith(flavor_const, "solaris") ||
+        StringStartsWith(flavor_const, "sunos"))
+    {
+        /* In this case minor is first and major is last.
+         * E.g. SunOS 5.11 where 11 is the major release number.
+         * Thus we need to swap them.
+         */
```
Comment (olehermanse) [inline libenv/sysinfo.c:3506]:
```suggestion
        /* In this case the major version comes last (second).
         * E.g. SunOS 5.11 where 11 is the major release number.
         * Thus we need to swap them.
         */
```

File: libenv/sysinfo.c [hunk 5]
```diff
@@ -3465,6 +3465,69 @@ static void SysOSNameHuman(EvalContext *ctx)
...
+        }
+        else if (minor == NULL && isdigit(*sp))
+        {
+            minor = sp;
+        }
+        else if (!isdigit(*sp))
+        {
+            *sp = '\0';
+        }
+    }
```
Comment (olehermanse) [inline libenv/sysinfo.c:3498]: This can be a separate function with unit tests.

File: libenv/sysinfo.c [hunk 6]
```diff
@@ -3465,6 +3465,69 @@ static void SysOSNameHuman(EvalContext *ctx)
...
+
+    char *major = NULL;
+    char *minor = NULL;
+    for (char *sp = flavor; *sp != '\0'; sp++)
+    {
+        if (major == NULL && isdigit(*sp))
+        {
+            major = sp;
+        }
+        else if (minor == NULL && isdigit(*sp))
```
Comment (olehermanse) [inline libenv/sysinfo.c:3490]: If major version has mutiple digits, minor will be assigned to the second digit of major.

In general, I think something like this is less error prone:

```C
while (*ptr != '\0' && !isdigit(*ptr))
{
    ++ptr; // Looking for first digit
}
if (*ptr == '\0')
{
    // Return early without assigning major
    // Or perhaps add an assert / abort / error here
    return;
}
major = ptr;
while (isdigit(*ptr))
{
    ++ptr; // Looking for end of first version number
}
if (*ptr == '.')
{
    *ptr = '\0';
    ptr++;
    if (isdigit(*ptr)) {
        minor = ptr;
        while (isdigit(*ptr)) ++ptr; // Looking for end of second version number (minor)
        *ptr = '\0':
    }
}
else
{
    // Assert that it ended at a zero byte, or do something else?
}
```

**Review by olehermanse — CHANGES_REQUESTED** [review]

File: libenv/sysinfo.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (vpodzime) [reply libenv/sysinfo.c:3506]: Yup, you are right, I didn't read the comment carefully enough.

File: libenv/sysinfo.c (hunk 6 above)
Comment (larsewi) [reply libenv/sysinfo.c:3490]: Good catch. Indeed it's safer to do this step by step.

**Comment by larsewi** [comment]
@cf-bottom Jenkins with exotics please !

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6472/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6472/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6472/

File: libenv/sysinfo.c [hunk 7]
```diff
@@ -3465,6 +3465,93 @@ static void SysOSNameHuman(EvalContext *ctx)
 
 /*****************************************************************************/
 
+/**
+ * Extract version number from string in place.
+ * 
+ * @param [in]  str string in which to extract version number from
```
Comment (vpodzime) [inline libenv/sysinfo.c:3471]: ~in which~

**Review by vpodzime — DISMISSED** [review]
Looks good to me otherwise. The failures in jenkins seems to be unrelated.

**Review by vpodzime — DISMISSED** [review]
ACK

File: libenv/sysinfo.c [hunk 8]
```diff
@@ -3465,6 +3465,93 @@ static void SysOSNameHuman(EvalContext *ctx)
...
+ * 
+ * @param [in]  str string to extract version number from
+ * @param [out] num pointer to start of version number or %NULL if no version
+ *                  number was found
+ * 
+ * @return pointer to the remaining string in `str` or %NULL if no remainder
+ * 
+ * @note `str` will be mutated
+ */
+static char *ExtractVersionNumber(char *str, char **num)
```
Comment (olehermanse) [inline libenv/sysinfo.c:3479]: This function does not have much to do with version numbers. (It doesn't even care about dots, dashes, or spaces).

Maybe call it `FindNextInteger()`?

File: tests/unit/sysinfo_test.c [hunk 9]
```diff
@@ -13,12 +14,56 @@ static void test_uptime(void)
...
+static void test_extract_version_number(void)
+{
+    char version_str[CF_BUFSIZE];
+    char *major;
+    char *minor;
+    char *patch;
+    char *bogus;
+    char *next;
+    
+    strcpy(version_str, "Ubuntu 20.04.1 LTS");
```
Comment (olehermanse) [inline tests/unit/sysinfo_test.c:26]: Missing some edge cases here:

```
""
" "
" no numbers in sight "
"0"
"1k"
"1234"
"1 2 3 4"
"Debian 9"
"Cent OS 6.7"
"CFEngine 3.18.0-2deadbeef"
```

Tips: You could make a wrapper for the test which calls the function in a loop, adding each number to a `Seq`. Then it's easier and less repetitive to test more cases.

**Review by olehermanse — CHANGES_REQUESTED** [review]
Great! Just some smaller comment. The only thing that needs to be changed is the commit message.

File: libenv/sysinfo.c [hunk 10]
```diff
@@ -3465,6 +3465,93 @@ static void SysOSNameHuman(EvalContext *ctx)
 
 /*****************************************************************************/
 
+/**
```
[the remark this answers is not in the document]
Comment (larsewi) [reply libenv/sysinfo.c:3468]: Good catch

**Comment by larsewi** [comment]
@cf-bottom Jenkins with exotics please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6481/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6481/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6481/

**Review by olehermanse — APPROVED** [review]
Great work!
